_studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_frame.cpp

```diff
@@ -0,0 +1,316 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "umc_defs.h"
+#if defined (MFX_ENABLE_MPEG2_VIDEO_DECODE)
+
+#include "umc_mpeg2_defs.h"
+#include "umc_mpeg2_slice.h"
+#include "umc_mpeg2_frame.h"
+
+namespace UMC_MPEG2_DECODER
+{
+    MPEG2DecoderFrameInfo::MPEG2DecoderFrameInfo(MPEG2DecoderFrame& Frame)
+        : frame(Frame)
+        , isField(false)
+        , isBottomField(false)
+        , frameType((FrameType)0)
+        , isFilled(false)
+        , sliceCount(0)
+    {
+    }
+
+    MPEG2DecoderFrameInfo::~MPEG2DecoderFrameInfo()
+    {
+        std::for_each(slices.begin(), slices.end(),
+            std::default_delete<MPEG2Slice>()
+        );
+    }
+
+    void MPEG2DecoderFrameInfo::Reset()
+    {
+        isField       = false;
+        isBottomField = false;
+        frameType     = (FrameType)0;
+        isFilled      = false;
+
+        std::for_each(slices.begin(), slices.end(),
+            std::default_delete<MPEG2Slice>()
+        );
+
+        slices.resize(0);
+        sliceCount = 0;
+        FreeReferenceFrames();
+
+    }
+
+    void MPEG2DecoderFrameInfo::AddSlice(MPEG2Slice* slice)
+    {
+        if (0 == sliceCount) // on the first slice
+        {
+            const auto pic    = slice->GetPicHeader();
+            const auto picExt = slice->GetPicExtHeader();
+            isField       =  picExt.picture_structure != FRM_PICTURE;
+            isBottomField = (picExt.picture_structure == BOTTOM_FLD_PICTURE);
+            frameType     = (FrameType)pic.picture_coding_type;
+        }
+
+        slices.push_back(slice);
+        sliceCount++;
+    }
+
+    void MPEG2DecoderFrameInfo::UpdateReferences(const DPBType & dpb)
+    {
+        const auto pic    = GetSlice(0)->GetPicHeader();
+        const auto picExt = GetSlice(0)->GetPicExtHeader();
+
+        if (MPEG2_I_PICTURE == pic.picture_coding_type)
+            return;
+
+        // 7.6.3.5 Prediction in P-pictures (for IP pair)
+        if (FRM_PICTURE != frame.pictureStructure)
+        {
+            uint8_t currFieldIndex = frame.GetNumberByParity(picExt.picture_structure == BOTTOM_FLD_PICTURE);
+            if ((1 == currFieldIndex) && (MPEG2_P_PICTURE == pic.picture_coding_type)) // current picture is P and the 2nd field
+            {
+                if (MPEG2_I_PICTURE == frame.GetAU(0)->GetSlice(0)->GetPicHeader().picture_coding_type) // the first field is I
+                {
+                    references.forward = &frame;
+                    references.forward->IncrementReference();
+                    return;
+                }
+            }
+        }
+
+        // 7.6.1 Frame/Field prediction
+        DPBType refFrames = dpb;
+        refFrames.remove_if(
```

Comment:
Can copy_if be used instead of full copy and remove_if?

Reply:
ok